perf-tool/src/methodEntry.cpp

```diff
@@ -36,54 +36,63 @@ using json = nlohmann::json;
 JNIEXPORT void JNICALL MethodEntry(jvmtiEnv *jvmtiEnv,
             JNIEnv* env,
             jthread thread,
             jmethodID method) {
 
-    if (mEntrySampleCount % mEntrySampleRate == 0) {
+    int numMethods;
+    /* Get number of methods and increment */
+    numMethods = atomic_fetch_add(&mEntrySampleCount, 1);
 
+    if (numMethods % mEntrySampleRate == 0) {           
         json j;
         jvmtiError err;
         char *name_ptr;
         char *signature_ptr;
-        char *generic_ptr;
         char *declaringClassName;
         jclass declaring_class;
         jint entry_count_ptr;
         jvmtiLineNumberEntry* table_ptr;
-        jlocation start_loc_ptr;
-        jlocation end_loc_ptr;
 
-        err = jvmtiEnv->GetMethodName(method, &name_ptr, &signature_ptr, &generic_ptr);
-        err = jvmtiEnv->GetMethodDeclaringClass(method, &declaring_class);
-        err = jvmtiEnv->GetClassSignature(declaring_class, &declaringClassName, NULL);
-        err = jvmtiEnv->GetMethodLocation(method, &start_loc_ptr, &end_loc_ptr);
-        err = jvmtiEnv->GetLineNumberTable(method, &entry_count_ptr, &table_ptr);
-
-        if (err == JVMTI_ERROR_NONE) {
+        j["methodNum"] = numMethods;
+        
+        err = jvmtiEnv->GetMethodName(method, &name_ptr, &signature_ptr, NULL);
+        if (check_jvmti_error(jvmtiEnv, err, "Unable to retrieve Method Name.\n")) {
             j["methodName"] = name_ptr;
             j["methodSig"] = signature_ptr;
-            j["methodClass"] = declaringClassName;
+        }
+        err = jvmtiEnv->GetMethodDeclaringClass(method, &declaring_class);
+        if (check_jvmti_error(jvmtiEnv, err, "Unable to retrieve Method Declaring Class.\n")) {
+            err = jvmtiEnv->GetClassSignature(declaring_class, &declaringClassName, NULL);
+            if (check_jvmti_error(jvmtiEnv, err, "Unable to retrieve Method Declaring Class Signature.\n")) {
+                j["methodClass"] = declaringClassName;
+            }
+        }
+        err = jvmtiEnv->GetLineNumberTable(method, &entry_count_ptr, &table_ptr);
+        if (check_jvmti_error(jvmtiEnv, err, "Unable to retrieve Method Line Number Table.\n")) {
             j["methodLineNum"] = table_ptr->line_number;
-            j["methodNum"] = mEntrySampleCount.load();
         }
 
         err = jvmtiEnv->Deallocate((unsigned char*)name_ptr);
+        check_jvmti_error(jvmtiEnv, err, "Unable to deallocate name_ptr.\n");
         err = jvmtiEnv->Deallocate((unsigned char*)signature_ptr);
-        err = jvmtiEnv->Deallocate((unsigned char*)generic_ptr);
+        check_jvmti_error(jvmtiEnv, err, "Unable to deallocate signature_ptr.\n");
         err = jvmtiEnv->Deallocate((unsigned char*)declaringClassName);
-
+        check_jvmti_error(jvmtiEnv, err, "Unable to deallocate declaringClassName.\n");
+        err = jvmtiEnv->Deallocate((unsigned char*)table_ptr);
+        check_jvmti_error(jvmtiEnv, err, "Unable to deallocate table_ptr.\n");
+    

```

Comment:
These deallocates need to stay as close as possible to the source of the allocation. If JVMTI call fails, then the strings won't be available to be deallocated.